File: internal/infrastructure/persistence/role.go
package persistence

import (
	"context"
	"sort"
	"sync"

	"go-clean-template/internal/domain"
)

type RoleRepository struct {
	mu    sync.RWMutex
	roles map[string]*domain.Role
}

func NewRoleRepository() *RoleRepository {
	return &RoleRepository{
		roles: make(map[string]*domain.Role),
	}
}

func (r *RoleRepository) FindByID(_ context.Context, id string) (*domain.Role, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	role, ok := r.roles[id]
	if !ok {
		return nil, domain.ErrNotFound
	}
	return role, nil
}

// FindAll returns all roles sorted by CreatedAt ascending, then applies
// offset/limit. The deterministic sort ensures stable pagination across calls.
func (r *RoleRepository) FindAll(_ context.Context, offset, limit int) ([]*domain.Role, int64, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()

	all := make([]*domain.Role, 0, len(r.roles))
	for _, role := range r.roles {
		all = append(all, role)
	}
	sort.Slice(all, func(i, j int) bool {
		return all[i].CreatedAt.Before(all[j].CreatedAt)
	})

	total := int64(len(all))
	if offset >= len(all) {
		return nil, total, nil
	}
	end := offset + limit
	if end > len(all) {
		end = len(all)
	}
	return all[offset:end], total, nil
}

func (r *RoleRepository) FindByName(_ context.Context, name string) (*domain.Role, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	for _, role := range r.roles {
		if role.Name == name {
			return role, nil
		}
	}
	return nil, domain.ErrNotFound
}

func (r *RoleRepository) Create(_ context.Context, role *domain.Role) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.roles[role.ID] = role
	return nil
}

func (r *RoleRepository) Update(_ context.Context, role *domain.Role) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	if _, ok := r.roles[role.ID]; !ok {
		return domain.ErrNotFound
	}
	r.roles[role.ID] = role
	return nil
}

func (r *RoleRepository) Delete(_ context.Context, id string) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	if _, ok := r.roles[id]; !ok {
		return domain.ErrNotFound
	}
	delete(r.roles, id)
	return nil
}

File: src/infrastructure/persistence/role.rs
use std::collections::HashMap;
use std::sync::RwLock;

use async_trait::async_trait;

use crate::application::usecase::user::RoleRepository;
use crate::domain;

pub struct InMemoryRoleRepository {
    roles: RwLock<HashMap<String, domain::Role>>,
}

impl InMemoryRoleRepository {
    pub fn new() -> Self {
        InMemoryRoleRepository {
            roles: RwLock::new(HashMap::new()),
        }
    }
}

impl Default for InMemoryRoleRepository {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl RoleRepository for InMemoryRoleRepository {
    async fn find_by_id(&self, id: &str) -> Result<domain::Role, domain::Error> {
        let roles = self.roles.read().map_err(|_| {
            domain::Error::InvalidInput("lock poisoned".to_string())
        })?;
        match roles.get(id) {
            Some(role) => Ok(role.clone()),
            None => Err(domain::Error::NotFound),
        }
    }

    async fn find_all(
        &self,
        offset: usize,
        limit: usize,
    ) -> Result<(Vec<domain::Role>, i64), domain::Error> {
        let roles = self.roles.read().map_err(|_| {
            domain::Error::InvalidInput("lock poisoned".to_string())
        })?;
        let mut all: Vec<&domain::Role> = roles.values().collect();
        all.sort_by_key(|a| a.created_at);

        let total = all.len() as i64;
        let slice: Vec<domain::Role> = all
            .into_iter()
            .skip(offset)
            .take(limit)
            .cloned()
            .collect();

        Ok((slice, total))
    }

    async fn find_by_name(&self, name: &str) -> Result<domain::Role, domain::Error> {
        let roles = self.roles.read().map_err(|_| {
            domain::Error::InvalidInput("lock poisoned".to_string())
        })?;
        for role in roles.values() {
            if role.name == name {
                return Ok(role.clone());
            }
        }
        Err(domain::Error::NotFound)
    }

    async fn create(&self, role: &domain::Role) -> Result<(), domain::Error> {
        let mut roles = self.roles.write().map_err(|_| {
            domain::Error::InvalidInput("lock poisoned".to_string())
        })?;
        roles.insert(role.id.clone(), role.clone());
        Ok(())
    }

    async fn update(&self, role: &domain::Role) -> Result<(), domain::Error> {
        let mut roles = self.roles.write().map_err(|_| {
            domain::Error::InvalidInput("lock poisoned".to_string())
        })?;
        if !roles.contains_key(&role.id) {
            return Err(domain::Error::NotFound);
        }
        roles.insert(role.id.clone(), role.clone());
        Ok(())
    }

    async fn delete(&self, id: &str) -> Result<(), domain::Error> {
        let mut roles = self.roles.write().map_err(|_| {
            domain::Error::InvalidInput("lock poisoned".to_string())
        })?;
        if roles.remove(id).is_none() {
            return Err(domain::Error::NotFound);
        }
        Ok(())
    }
}
